Python/Product/Analysis/SourceSpan.cs

```diff
@@ -84,8 +75,20 @@ public SourceSpan(int startLine, int startColumn, int endLine, int endColumn)
         /// <summary>
         /// Whether the locations in the span are valid.
         /// </summary>
-        public bool IsValid {
-            get { return _start.IsValid && _end.IsValid; }
+        public bool IsValid => Start.IsValid && End.IsValid;
+
+        public SourceSpan Union(SourceSpan other) {
+            if (other.Start.Line < Start.Line) {
+                Start = other.Start;
+            } else if(other.Start.Line == Start.Line && other.Start.Column < Start.Column) {
+                Start = other.Start;
+            }
+            if (other.End.Line > End.Line) {
+                End = other.End;
+            } else if (other.End.Line == End.Line && other.End.Column > End.Column) {
+                End = other.End;
+            }
+            return this;
```

Comment:
Let's not make mutable structs. We can return new instance of `SourceSpan` here.

Reply:
OK